File: src/scala/implementation/InnerSet_9.scala
package scala.implementation

object InnerSet_9
{
    def apply[T](element1: T, element2: T, element3: T, element4: T, element5: T, element6: T, element7: T, element8: T, element9: T): InnerSet_9[T] = new InnerSet_9[T](element1, element2, element3, element4, element5, element6, element7, element8, element9)
}

class InnerSet_9[T](element1: T, element2: T, element3: T, element4: T, element5: T, element6: T, element7: T, element8: T, element9: T) extends AbstractInnerSet[T]
{

    override def addElement(p_element: T): InnerSet[T] =
    {
        if (element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element) || element5.equals(p_element) || element6.equals(p_element) || element7.equals(p_element) || element8.equals(p_element) || element9.equals(p_element))
        {
            this
        }
        else  InnerSet_10(element1, element2, element3, element4, element5, element6, element7, element8, element9, p_element)
    }

    override def addUnchecked(p_element: T): InnerSet[T] = InnerSet_10(element1, element2, element3, element4, element5, element6, element7, element8, element9, p_element)


    override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =     {
                p_innerSet.addElement(element1)
                p_innerSet.addElement(element2)
                p_innerSet.addElement(element3)
                p_innerSet.addElement(element4)
                p_innerSet.addElement(element5)
                p_innerSet.addElement(element6)
                p_innerSet.addElement(element7)
                p_innerSet.addElement(element8)
                p_innerSet.addElement(element9)
            }
    override def containsElements(p_element: T): Boolean = element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element) || element5.equals(p_element) || element6.equals(p_element) || element7.equals(p_element) || element8.equals(p_element) || element9.equals(p_element)

    override def containsAllElements(p_innerSet: InnerSet[T]): Boolean =  p_innerSet.containsElements(element1) && p_innerSet.containsElements(element2) && p_innerSet.containsElements(element3) && p_innerSet.containsElements(element4) && p_innerSet.containsElements(element5) && p_innerSet.containsElements(element6) && p_innerSet.containsElements(element7) && p_innerSet.containsElements(element8) && p_innerSet.containsElements(element9)

    override def getElement(p_index: Int): Option[T] = p_index match{
        case 1 => Option(element1)
        case 2 => Option(element2)
        case 3 => Option(element3)
        case 4 => Option(element4)
        case 5 => Option(element5)
        case 6 => Option(element6)
        case 7 => Option(element7)
        case 8 => Option(element8)
        case 9 => Option(element9)
        case _ => None
    }


    override def removeElement(p_element: T): InnerSet[T] =
    {
            if(element1.equals(p_element))
         InnerSet_8(element2, element3, element4, element5, element6, element7, element8, element9)
                  else             if(element2.equals(p_element))
         InnerSet_8(element1, element3, element4, element5, element6, element7, element8, element9)
                  else             if(element3.equals(p_element))
         InnerSet_8(element1, element2, element4, element5, element6, element7, element8, element9)
                  else             if(element4.equals(p_element))
         InnerSet_8(element1, element2, element3, element5, element6, element7, element8, element9)
                  else             if(element5.equals(p_element))
         InnerSet_8(element1, element2, element3, element4, element6, element7, element8, element9)
                  else             if(element6.equals(p_element))
         InnerSet_8(element1, element2, element3, element4, element5, element7, element8, element9)
                  else             if(element7.equals(p_element))
         InnerSet_8(element1, element2, element3, element4, element5, element6, element8, element9)
                  else             if(element8.equals(p_element))
         InnerSet_8(element1, element2, element3, element4, element5, element6, element7, element9)
                  else             if(element9.equals(p_element))
         InnerSet_8(element1, element2, element3, element4, element5, element6, element7, element8)
                            else this
    }


    override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =     {
                p_innerSet.removeElement(element1)
                p_innerSet.removeElement(element2)
                p_innerSet.removeElement(element3)
                p_innerSet.removeElement(element4)
                p_innerSet.removeElement(element5)
                p_innerSet.removeElement(element6)
                p_innerSet.removeElement(element7)
                p_innerSet.removeElement(element8)
                p_innerSet.removeElement(element9)
            }

    override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
    {
      var i: InnerSet[T] = this
            if(!p_innerSet.containsElements(element1)) i = i.removeElement(element1)
            if(!p_innerSet.containsElements(element2)) i = i.removeElement(element2)
            if(!p_innerSet.containsElements(element3)) i = i.removeElement(element3)
            if(!p_innerSet.containsElements(element4)) i = i.removeElement(element4)
            if(!p_innerSet.containsElements(element5)) i = i.removeElement(element5)
            if(!p_innerSet.containsElements(element6)) i = i.removeElement(element6)
            if(!p_innerSet.containsElements(element7)) i = i.removeElement(element7)
            if(!p_innerSet.containsElements(element8)) i = i.removeElement(element8)
            if(!p_innerSet.containsElements(element9)) i = i.removeElement(element9)
            i
    }

    override def getSize: Int = 9

    override def iterator: InnerSetIterator[T] = InnerSetIterator(this)

    override def copy: Option[InnerSet[T]] = Option(InnerSet_9(element1: T, element2: T, element3: T, element4: T, element5: T, element6: T, element7: T, element8: T, element9: T))

    override def clear(unused: Boolean): InnerSet[T] = new InnerSet_0

    override def toString : String = //TODO
}

File: src/j/implementation/microSet/InnerSet_9.java
package j.implementation.microSet;

import java.util.Iterator;

public class InnerSet_9 <T> extends AbstractInnerSet<T>
{
        protected T element_1;
        protected T element_2;
        protected T element_3;
        protected T element_4;
        protected T element_5;
        protected T element_6;
        protected T element_7;
        protected T element_8;
        protected T element_9;
        
    public InnerSet_9(T p_element_1, T p_element_2, T p_element_3, T p_element_4, T p_element_5, T p_element_6, T p_element_7, T p_element_8, T p_element_9)
    {
                element_1 = p_element_1;
                element_2 = p_element_2;
                element_3 = p_element_3;
                element_4 = p_element_4;
                element_5 = p_element_5;
                element_6 = p_element_6;
                element_7 = p_element_7;
                element_8 = p_element_8;
                element_9 = p_element_9;
            }

    public InnerSet_9(InnerSet_9<T> i)
    {
            element_1 = i.element_1;
            element_2 = i.element_2;
            element_3 = i.element_3;
            element_4 = i.element_4;
            element_5 = i.element_5;
            element_6 = i.element_6;
            element_7 = i.element_7;
            element_8 = i.element_8;
            element_9 = i.element_9;
        }

    public InnerSet_9<T> copy() {
        return this;
    }

    public InnerSet_9(InnerSet_8<T> i, T el)
    {
            element_1 = i.element_1;            element_2 = i.element_2;            element_3 = i.element_3;            element_4 = i.element_4;            element_5 = i.element_5;            element_6 = i.element_6;            element_7 = i.element_7;            element_8 = i.element_8;                    element_9 = el;
    }
    
    @Override
    public InnerSet<T> addElement(T p_element)
    {
        if(element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element) || element_4.equals(p_element) || element_5.equals(p_element) || element_6.equals(p_element) || element_7.equals(p_element) || element_8.equals(p_element) || element_9.equals(p_element))
            return this;
        else
            if (shouldArray())
                return new InnerArraySet<T>(this, p_element);
            else
                return new InnerSet_10<>(this, p_element);    }

    @Override
    public InnerSet<T> addUnChecked(T p_element)
    {
        if (shouldArray())
            return new InnerArraySet<T>(this, p_element);
        else
            return new InnerSet_10<>(this, p_element);    }

    @Override
    public InnerSet<T> removeElement(Object p_element)
    {
                if(element_1.equals(p_element))
            return  new InnerSet_8<>(element_2, element_3, element_4, element_5, element_6, element_7, element_8, element_9);
         else                 if(element_2.equals(p_element))
            return  new InnerSet_8<>(element_1, element_3, element_4, element_5, element_6, element_7, element_8, element_9);
         else                 if(element_3.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_4, element_5, element_6, element_7, element_8, element_9);
         else                 if(element_4.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_3, element_5, element_6, element_7, element_8, element_9);
         else                 if(element_5.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_3, element_4, element_6, element_7, element_8, element_9);
         else                 if(element_6.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_3, element_4, element_5, element_7, element_8, element_9);
         else                 if(element_7.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_3, element_4, element_5, element_6, element_8, element_9);
         else                 if(element_8.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_3, element_4, element_5, element_6, element_7, element_9);
         else                 if(element_9.equals(p_element))
            return  new InnerSet_8<>(element_1, element_2, element_3, element_4, element_5, element_6, element_7, element_8);
                        else
            return this;
    }

    @Override
    public boolean containsElement(Object p_element)
    {
        return element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element) || element_4.equals(p_element) || element_5.equals(p_element) || element_6.equals(p_element) || element_7.equals(p_element) || element_8.equals(p_element) || element_9.equals(p_element);
    }


    public T getElement(int index)
    {switch(index) {
                case 0:
            return element_1;
                case 1:
            return element_2;
                case 2:
            return element_3;
                case 3:
            return element_4;
                case 4:
            return element_5;
                case 5:
            return element_6;
                case 6:
            return element_7;
                case 7:
            return element_8;
                case 8:
            return element_9;
                default:
            return null;
            }
    }


    public InnerSet<T> addAllElements(InnerSet<T> innerSet) {
        // Reverse the add if the innerSet we add is smaller than this
        if (innerSet.getSize() < this.getSize()) {
            return innerSet.addAllElements(this);
        }
        return innerSet.addElement(element_1).addElement(element_2).addElement(element_3).addElement(element_4).addElement(element_5).addElement(element_6).addElement(element_7).addElement(element_8).addElement(element_9);
    }


    @Override
    public InnerSet<T> removeAllElements(InnerSet<T> innerSet) {
        return innerSet.removeElement(element_1).removeElement(element_2).removeElement(element_3).removeElement(element_4).removeElement(element_5).removeElement(element_6).removeElement(element_7).removeElement(element_8).removeElement(element_9);
    }

    @Override
    public InnerSet<T> retainAllElements(InnerSet<T> innerSet) {
        InnerSet<T> i = this;
                if (!innerSet.containsElement(element_1)) i = i.removeElement(element_1);
                if (!innerSet.containsElement(element_2)) i = i.removeElement(element_2);
                if (!innerSet.containsElement(element_3)) i = i.removeElement(element_3);
                if (!innerSet.containsElement(element_4)) i = i.removeElement(element_4);
                if (!innerSet.containsElement(element_5)) i = i.removeElement(element_5);
                if (!innerSet.containsElement(element_6)) i = i.removeElement(element_6);
                if (!innerSet.containsElement(element_7)) i = i.removeElement(element_7);
                if (!innerSet.containsElement(element_8)) i = i.removeElement(element_8);
                if (!innerSet.containsElement(element_9)) i = i.removeElement(element_9);
                return i;
    }

    @Override
    public boolean containsAllElements(InnerSet<T> innerSet) {
        return  innerSet.containsElement(element_1) && innerSet.containsElement(element_2) && innerSet.containsElement(element_3) && innerSet.containsElement(element_4) && innerSet.containsElement(element_5) && innerSet.containsElement(element_6) && innerSet.containsElement(element_7) && innerSet.containsElement(element_8) && innerSet.containsElement(element_9);
    }

    public InnerSet<T> addAllAndPropagate(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAllAndPropagateReverse(this, microSetToPropagate);
    }


    public InnerSet<T> addAllAndPropagateReverse(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAndPropagate(element_1, microSetToPropagate).addAndPropagate(element_2, microSetToPropagate).addAndPropagate(element_3, microSetToPropagate).addAndPropagate(element_4, microSetToPropagate).addAndPropagate(element_5, microSetToPropagate).addAndPropagate(element_6, microSetToPropagate).addAndPropagate(element_7, microSetToPropagate).addAndPropagate(element_8, microSetToPropagate).addAndPropagate(element_9, microSetToPropagate);
    }

    public InnerSet<T> addAndPropagate(T p_element, MicroSet<T> microSetToPropagate) {
        if(element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element) || element_4.equals(p_element) || element_5.equals(p_element) || element_6.equals(p_element) || element_7.equals(p_element) || element_8.equals(p_element) || element_9.equals(p_element))
            return this;
        else {
            microSetToPropagate.add(p_element);
            if (shouldArray())
                return new InnerArraySet<T>(this, p_element);
            else
                return new InnerSet_10<>(this, p_element);        }
    }

    @Override
    public String toString()
    {
        return "{" + element_1 + ", " + element_2 + ", " + element_3 + ", " + element_4 + ", " + element_5 + ", " + element_6 + ", " + element_7 + ", " + element_8 + ", " + element_9 + " }";
    }
    
    @Override
    public int getSize()
    {
        return 9;
    }

    @Override
    public Iterator<T> iterator() {
        return new InnerSetIterator<>(this);
    }

    @Override
    public InnerSet<T> clear(boolean unused) {
        return (InnerSet<T>) InnerSet_0.singleton;
    }
}
